src/formats/cif.cpp

```diff
@@ -0,0 +1,205 @@
+#include <string>
+#include <vector>
+#include <stdexcept>
+#include <fstream>
+#include <filesystem>
+
+#include <gemmi/cif.hpp>
+#include <gemmi/mmcif.hpp>
+
+#define GEMMI_WRITE_IMPLEMENTATION
+#include <gemmi/to_cif.hpp>
+#undef GEMMI_WRITE_IMPLEMENTATION
+
+#include "chargefw2.h"
+#include "cif.h"
+#include "../structures/molecule_set.h"
+#include "../charges.h"
+#include "../config.h"
+
+
+namespace fs = std::filesystem;
+
+static const std::vector<std::string> atom_site_columns{
+        "group_PDB",
+        "auth_asym_id",
+        "auth_seq_id",
+        "pdbx_PDB_ins_code",
+        "auth_comp_id",
+        "auth_atom_id",
+        "label_alt_id",
+        "type_symbol",
+        "Cartn_x",
+        "Cartn_y",
+        "Cartn_z",
+        "pdbx_PDB_model_num",
+    };
+
+
+class MCRA {
+    const int _model;
+    const std::string _chain;
+    const std::string _res_num;
+    const std::string _residue;
+    const std::string _atom;
+
+public:
+    MCRA(const int model,
+         const std::string &chain,
+         const std::string &res_num,
+         const std::string &residue,
+         const std::string &atom)
+        : _model(model), _chain(chain), _res_num(res_num), _residue(residue), _atom(atom)
+        {}
+
+    int find_row(gemmi::cif::Table &table, int start_idx = 0) const {
+        int idx = start_idx; 
+        int total_rows = table.length();
+
+        assert(idx < total_rows);
+
+        while (idx < total_rows) {
+            if (this->is_row(table[idx])) {
+                return idx;
+            }
+            ++idx;
+        }
+
+        fmt::print(stderr, "In MCRA::find_row(), unable to find Atom(1, {}, {}, {}, {})\n"
+                    "Looping through entire table to double check...",
+                    this->_chain, this->_res_num, this->_residue, this->_atom); 
+
+        idx = 0;
+        for (const auto site : table) {
+            if (this->is_row(site)) {
+                return idx;
+            }
+            ++idx;
+        }
+        return -1;
+    }
+
+    friend std::ostream &operator<<(std::ostream &os, const MCRA &mcra) {
+        os << "Atom(" << mcra._model << " " << mcra._chain << " " << mcra._res_num << " [" << mcra._residue << "] " << mcra._atom << ")";
+        return os;
+    }
+
+private:
+    bool is_row(const gemmi::cif::Table::Row &row) const {
+        int model = std::stoul(row[11]);
+        const std::string &chain = row[1];
+        const std::string &residue = row[4];
+        const std::string &res_num = row[2];
+        const std::string atom = row[5][0] != '"' ? row[5] : row[5].substr(1, row[5].size() - 2);
+
+        if (_model == model) {
+            if (_chain == chain) {
+                if (_res_num == res_num && _residue == residue) {
+                    if (_atom == atom) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+};
+
+
+void CIF::write_cif_block(std::ostream &out,
```

Comment:
I believe that the output CIF file should be again directly usable as an input for ChargeFW2 - right now the new tags are added again, resulting in GEMMI not reading the file correctly:

```
doc = gemmi.cif.read_file('/tmp/1rud.fw2.fw2.cif') 
RuntimeError: /tmp/1rud.fw2.fw2.cif:1592 in data_1RUD: duplicate tag _atom_site.fw2_charge
```

Maybe overwrite the existing values?

Reply:
This is a nice catch! I never thought about this and is probably a bug also in freesasa.